### PR TITLE
dpl-712-mbrave-download

### PR DESCRIPTION
**Pull request by yoldas**
Closes #

Changes proposed in this pull request:

*
*
* ...

### CODE REVIEW

**Comment by yoldas** [comment]
See the following to be able to pass fields as string 

https://github.com/JsonApiClient/json_api_client/blob/4277a862edf2f80a168c2c7e0206325a9eda5a4a/lib/json_api_client/query/builder.rb#L222

**Comment by harrietc52** [comment]
Removing from Project board as referenced in EPIC PR [here](https://github.com/sanger/limber/issues/1270)

File: app/views/exports/bioscan_mbrave.tsv.erb [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+    first_tag_1_group_name = aliquot.tag.tag_group.name
+    version = mbrave_tag_version(first_tag_1_group_name)
+
+    rows = [['Forward Labels', 'Reverse Labels', 'Label', 'Group', 'UMI plate ID', 'Sample Plate ID']]
+    @tube.aliquots.each do |aliquot| 
+      tag_1_group_name = aliquot.tag.tag_group.name
+      tag_2_group_name = aliquot.tag2.tag_group.name
+
+      if (version != mbrave_tag_version(tag_1_group_name)) || (version != mbrave_tag_version(tag_2_group_name))
+        raise "This tube contains tags from different tag groups"
```
Comment (andrewsparkes) [inline app/views/exports/bioscan_mbrave.tsv.erb:12]: ..'from different tag group versions'

(the tube will always contain tags from different tag groups)

File: config/mbrave.yml [hunk 2]
```diff
@@ -0,0 +1,206 @@
+---
+default: &DEFAULT
+  bioscan_p1:
+    version: 1
```
Comment (andrewsparkes) [inline config/mbrave.yml:4]: there is already a version 2, so maybe we should start with that version (see Scotts slack message about what changed)

File: config/purposes/bespoke.yml [hunk 3]
```diff
@@ -31,6 +31,10 @@ LBB Lib Pool Stock:
   :creator_class: LabwareCreators::PooledTubesBySubmission
   :presenter_class: Presenters::SimpleTubePresenter
   :default_printer_type: :tube
+  :file_links:
+    - name: 'Download MBRAVE file'
```
Comment (andrewsparkes) [inline config/purposes/bespoke.yml:35]: I think it more accurately it is called MBRAVE UMI file (I think there are 3 MBrave input files altogether, this is one of them, there is also the sequencing data file and some sort of reference file)

File: spec/models/presenters/tube_presenter_spec.rb [hunk 4]
```diff
@@ -114,4 +114,19 @@
...
     end
   end
+  describe '#csv_links_for' do
+    let!(:purpose_config) { create(:tube_with_file_links_config, uuid: purpose_uuid) }
+    context 'when the file is a .tsv' do
+      it 'renders the right links' do
+        expect(subject.csv_file_links).to eq(
+          [
+            [
+              'Download MBRAVE file',
```
Comment (andrewsparkes) [inline spec/models/presenters/tube_presenter_spec.rb:124]: MBRAVE UMI file

**Review by andrewsparkes — APPROVED** [review]

File: app/views/exports/bioscan_mbrave.tsv.erb (hunk 1 above)
Comment (emrojo) [reply app/views/exports/bioscan_mbrave.tsv.erb:12]: I thought it was going to be 2 tag groups: tag_1 and tag_2, and then we check that all the 9600 aliquots are using the same tag groups (the same version).

File: config/purposes/bespoke.yml (hunk 3 above)
Comment (emrojo) [reply config/purposes/bespoke.yml:35]: My mistake, this is from when we were coding this before having the pipeline

File: config/mbrave.yml (hunk 2 above)
Comment (yoldas) [reply config/mbrave.yml]: We are starting with version 2 now. The tag name changes for version 2 have been added. Reverse grouping is has been merged.

File: config/purposes/bespoke.yml (hunk 3 above)
Comment (yoldas) [reply config/purposes/bespoke.yml]: This section is in config/purposes/bioscan.yml , under `LBSN-9216 Lib PCR Pool XP` purpose now.

File: spec/models/presenters/tube_presenter_spec.rb (hunk 4 above)
Comment (yoldas) [reply spec/models/presenters/tube_presenter_spec.rb]: Changed to Download MBRAVE file

File: config/purposes/bioscan.yml [hunk 5]
```diff
@@ -83,6 +85,7 @@ LBSN-384 PCR 2 Pool:
   :target: StockMultiplexedLibraryTube
   :type: IlluminaHtp::StockTubePurpose
   :creator_class: LabwareCreators::PooledTubesFromWholePlates
+  :transfer_template: '384 plate to tube'
```
Comment (andrewsparkes) [inline config/purposes/bioscan.yml:88]: What does this transfer_template parameter do? Curious as it isn't on the epic branch yet the pooling seems to work. Are you just being more explicit?

File: spec/factories/purpose_config_factories.rb [hunk 6]
```diff
@@ -53,6 +53,13 @@
       transfer_parameters { { target_molarity_nm: 4, target_volume_ul: 192, minimum_pick_ul: 2 } }
     end
 
+    # tube with file links
+    factory :tube_with_file_links_config do
+      name { 'Bioscan Pool Tube' }
+      file_links { [{ name: 'Download MBRAVE UMI file', id: 'bioscan_mbrave', format: 'tsv' }] }
+      presenter_class { 'Presenters::TubePresenter' }
```
Comment (andrewsparkes) [inline spec/factories/purpose_config_factories.rb:60]: Should this be FinalTubePresenter? or not important for this factory?

**Review by andrewsparkes — APPROVED** [review]
Looks good, just had some minor comments/questions but it can be merged I think

File: config/purposes/bioscan.yml (hunk 5 above)
Comment (emrojo) [reply config/purposes/bioscan.yml:88]: Is a reference to the transfer template in SS. it is defined there: <https://github.com/sanger/sequencescape/pull/3780/files#diff-36a74a283e00b28f937853568ea928e403aed96cd63b5c9b8d7b05f5632108cb>

File: spec/models/presenters/tube_presenter_spec.rb (hunk 4 above)
Comment (yoldas) [reply spec/models/presenters/tube_presenter_spec.rb]: Sorry, I meant to say 
Changed to `Download MBRAVE UMI file`

Should we change it to `Download MBrave UMI file` ? (Note the capital letters)

**Pull request merged**